source/retryable-reads/retryable-reads.rst

```diff
@@ -357,16 +358,22 @@ and reflects the flow described above.
      * can then assume that no attempts were made. */
     server = selectServer();
 
-    /* PoolClearedException indicates the operation did not even attempt to
-     * create a connection, let alone execute the operation. This means we
-     * are always safe to attempt a retry. We do not need to update SDAM,
-     * since whatever error caused the pool to be cleared will do so itself. */
     try {
       connection = server.getConnection()
     } catch (PoolClearedException poolClearedError) {
+      /* PoolClearedException indicates the operation did not even attempt to
+       * create a connection, let alone execute the operation. This means we
+       * are always safe to attempt a retry. We do not need to update SDAM,
+       * since whatever error caused the pool to be cleared will do so itself. */
+      return executeRetry(command, session, poolClearedError);
+    } catch (NetworkError networkError) {
+      /* NetworkError indicates the getConnection was not able to establish
+       * a successfully handshake-ed connection with the server, this is retryable.
+       * We do not need to update SDAM, since network errors already mark
+       * servers Unknown and empty the pool. */
```

Comment:
New thread because Github is not letting me reply on the existing one. Since the executeCommand case below includes updateTopologyDescriptionForNetworkError+updateTopologyDescriptionForNotWritablePrimaryError I think we should do the same here:
```
    try {
      connection = server.getConnection()
    } catch (PoolClearedException poolClearedError) {
      /* PoolClearedException indicates the operation did not even attempt to
       * create a connection, let alone execute the operation. This means we
       * are always safe to attempt a retry. We do not need to update SDAM,
       * since whatever error caused the pool to be cleared will do so itself. */
      return executeRetry(command, session, poolClearedError);
    } catch (NetworkError networkError) {
      updateTopologyDescriptionForNetworkError(server, networkError);
      return executeRetry(command, session, poolClearedError);
    } catch (NotWritablePrimaryException originalError) {
      updateTopologyDescriptionForNotWritablePrimaryError(server, originalError);
      return executeRetry(command, session, poolClearedError);
    }
```

Speaking of, is it possible to add test coverage for a NotWritablePrimaryException? Right now I think we're only testing NetworkError.

Reply:
Updaeted the code. Yes adding that coverage sounds reasonable, I'll post something Monday

Reply:
Pushed tests for NotWritablePrimary errors, I was able to test these against node and they pass 🚀